apis/core/v1beta1/openstackcontrolplane_webhook.go

```diff
@@ -46,14 +49,42 @@ var _ webhook.Validator = &OpenStackControlPlane{}
 func (r *OpenStackControlPlane) ValidateCreate() error {
 	openstackcontrolplanelog.Info("validate create", "name", r.Name)
 
-	return r.ValidateServices()
+	var allErrs field.ErrorList
+
+	allErrs = append(allErrs, r.ValidateServices()...)
+
+	if len(allErrs) != 0 {
+		return apierrors.NewInvalid(
```

Comment:
Should we use newInvalid() or newConflict()/newApplyConflict() instead?  It's not invalid value per se.

Reply:
Using the list, the idea is that the apierrors raised will be from all validations, including validations in individual operators. For example here we append errors from ironic-operator validator - https://github.com/openstack-k8s-operators/openstack-operator/blob/92bcbf201c3bb6b399fc1b605f80cedaae185904/apis/core/v1beta1/openstackcontrolplane_webhook.go#L54-L67

Each error in "allErrs" can be of different type, conflict, forbidden, invalid etc.
 I think "Invalid" for each dependency check might not be the best error, but I think we may want to iterate and add some custom error types?

At this level a NewInvalid seems appropriate? It is the user input that is invalid?
But for individual errors in "allErrs" we can set individual types.

Reply:
The way newInvaid() is rendered in the response[1]  can be confusing when it says the field is invalid. "it's valid field value but it conflicts with other values". Probably having our own custom StatusError would be better.


[1] https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/webhook/admission/validator.go#L70
[2] https://github.com/kubernetes/apimachinery/blob/master/pkg/api/errors/errors.go#L145

Reply:
This is what kubectr/oc outputs dependency errors in multiple services:

```
The OpenstackControlplane "openstack" is invalid: 
* spec.Keystone.enabled: Invalid value: true: Keystone service dependencies are not enabled, needs: Mariadb
* spec.Neutron.enabled: Invalid value: true: Neutron service dependencies are not enabled, needs: Mariadb, Rabbitmq and Keystone
* spec.Nova.enabled: Invalid value: true: Nova service dependencies are not enabled, needs: Mariadb, Rabbitmq, Keystone, Placement, Neutron and Glance
* spec.Ironic.enabled: Invalid value: true: Ironic service dependencies are not enabled, needs: Mariadb and Keystone
make: *** [Makefile:280: openstack_deploy] Error 1
```

I think `The OpenstackControlplane "openstack" is invalid` is not bad, but each individual field we `Invalid value` is not so good, I don't se a better alternative in the package i.e would be good to follow up with custom error - MissingDependency or something?